### PR TITLE
Implement Managed Streaming for Kafka (MSK) - Cluster

### PR DESCRIPTION
**Pull request by haarchri**
Signed-off-by: Christopher Haar <chhaar30@googlemail.com>

### Description of your changes

kafka cluster  (MSK)
Implements #415

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

```
kubectl apply -f examples/kafka/cluster.yaml
cluster.kafka.aws.crossplane.io/kafka-test-server

```

```
kubectl get cluster.kafka.aws.crossplane.io/kafka-test-server

apiVersion: kafka.aws.crossplane.io/v1alpha1
kind: Cluster
metadata:
  annotations:
    crossplane.io/external-name: arn:aws:kafka:eu-central-1:255932642927:cluster/kafka-test-server/af7e1917-25f4-4418-861e-fa401d8beba4-1
  creationTimestamp: "2021-06-21T07:45:52Z"
  finalizers:
  - finalizer.managedresource.crossplane.io
  name: kafka-test-server
spec:
  deletionPolicy: Delete
  forProvider:
    brokerNodeGroupInfo:
      brokerAZDistribution: DEFAULT
      clientSubnets:
      - subnet-f2ccb898
      - subnet-aac361d6
      instanceType: kafka.t3.small
      securityGroups:
      - sg-7f714d0e
      storageInfo:
        ebsStorageInfo:
          volumeSize: 1
    clusterName: kafka-test-server
    encryptionInfo:
      encryptionAtRest:
        dataVolumeKMSKeyID: arn:aws:kms:eu-central-1:255932642927:key/6d90627a-663e-426c-b5c4-63578d217388
      encryptionInTransit:
        clientBroker: TLS
        inCluster: true
    enhancedMonitoring: DEFAULT
    kafkaVersion: 2.6.1
    numberOfBrokerNodes: 2
    openMonitoring:
      prometheus:
        jmxExporter:
          enabledInBroker: false
        nodeExporter:
          enabledInBroker: false
    region: eu-central-1
    tags:
      myKey: myValue
    zookeeperConnectString: z-3.kafka-test-server.xxxxxx.c1.kafka.eu-central-1.amazonaws.com:2181,z-2.kafka-test-server.xxxxxx.c1.kafka.eu-central-1.amazonaws.com:2181,z-1.kafka-test-server.xxxxxx.c1.kafka.eu-central-1.amazonaws.com:2181
    zookeeperConnectStringTLS: z-3.kafka-test-server.xxxxxx.c1.kafka.eu-central-1.amazonaws.com:2182,z-2.kafka-test-server.xxxxxx.c1.kafka.eu-central-1.amazonaws.com:2182,z-1.kafka-test-server.xxxxxx.c1.kafka.eu-central-1.amazonaws.com:2182
  providerConfigRef:
    name: default
status:
  atProvider:
    clusterARN: arn:aws:kafka:eu-central-1:255932642927:cluster/kafka-test-server/af7e1917-25f4-4418-861e-fa401d8beba4-1
    state: ACTIVE
  conditions:
  - lastTransitionTime: "2021-06-21T07:46:00Z"
    reason: ReconcileSuccess
    status: "True"
    type: Synced
  - lastTransitionTime: "2021-06-21T08:11:05Z"
    reason: Available
    status: "True"
    type: Ready
```

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Review by hasheddan — COMMENTED** [review]
> during deletion i see in debug the following message - but cluster is removed successfully in k8s and in aws ( process takes 5-10 minutes)
Message: delete failed: failed to delete Cluster: BadRequestException: You can't delete cluster in DELETING state. - any ideas here ?

@haarchri one option here would be to ignore this error in your `postDelete` hook. Let me know if you have any questions! :)

File: examples/kafka/cluster.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  forProvider:
+    brokerNodeGroupInfo:
+      clientSubnets:
+        - subnet-f2ccb898
+        - subnet-aac361d6
+      instanceType: kafka.t3.small
+      storageInfo: 
+        ebsStorageInfo:
+          volumeSize: 1
+    clusterName: kafka-test-server
```
Comment (negz) [inline examples/kafka/cluster.yaml:16]: Is this the unique identifier of the MSK cluster? Should it be the external name?

https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md#external-name

File: examples/kafka/cluster.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    kafkaVersion: 2.6.1
+    numberOfBrokerNodes: 2
+    region: eu-central-1
+    tags:
+      myKey: myValue
+  providerConfigRef:
+    name: default
```
Comment (negz) [inline examples/kafka/cluster.yaml:23]: I _think_ we're exclusively using a `providerConfigRef` named `example` elsewhere. I'd prefer to be consistent.

File: pkg/controller/aws.go [hunk 3]
```diff
@@ -73,6 +73,7 @@ import (
 	"github.com/crossplane/provider-aws/pkg/controller/identity/iamuser"
 	"github.com/crossplane/provider-aws/pkg/controller/identity/iamuserpolicyattachment"
 	"github.com/crossplane/provider-aws/pkg/controller/identity/openidconnectprovider"
+	kafkaCluster "github.com/crossplane/provider-aws/pkg/controller/kafka/cluster"
```
Comment (negz) [inline pkg/controller/aws.go:76]: Perhaps just `kafkacluster` here? This alias is valid, but camel case package names aren't particularly idiomatic in Go. We've previously used package names like `dbcluster`, and `certificateauthority` for example.

https://blog.golang.org/package-names#TOC_2.

File: examples/kafka/cluster.yaml (hunk 1 above)
Comment (negz) [reply examples/kafka/cluster.yaml:16]: Answering my own question - the ARN is the unique identifier, so no this should not be the external name. :)

**Review by negz — COMMENTED** [review]
This looks good to me apart from a few nits and once we address the delete issues.

File: examples/kafka/cluster.yaml (hunk 1 above)
Comment (haarchri) [reply examples/kafka/cluster.yaml:16]: yes ;)

File: examples/kafka/cluster.yaml (hunk 2 above)
Comment (haarchri) [reply examples/kafka/cluster.yaml]: done - also changed region to us ;)

File: pkg/controller/aws.go (hunk 3 above)
Comment (haarchri) [reply pkg/controller/aws.go]: done

**Comment by haarchri** [comment]
@hasheddan  i will have a look in postDelete and awsErr.Code() - if i'am right =)

**Comment by haarchri** [comment]
> > during deletion i see in debug the following message - but cluster is removed successfully in k8s and in aws ( process takes 5-10 minutes)
> > Message: delete failed: failed to delete Cluster: BadRequestException: You can't delete cluster in DELETING state. - any ideas here ?
> 
> @haarchri one option here would be to ignore this error in your `postDelete` hook. Let me know if you have any questions! :)

can you have a look ? 
https://github.com/haarchri/provider-aws/blob/kafka/pkg/controller/kafka/cluster/setup.go#L71

thanks ;)

File: pkg/controller/kafka/cluster/setup.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	obj.ClusterArn = awsclients.String(meta.GetExternalName(cr))
+	return false, nil
+}
+
+func postDelete(_ context.Context, cr *svcapitypes.Cluster, obj *svcsdk.DeleteClusterOutput, err error) error {
+	if err != nil {
+		// skip: err.Error(): failed to delete Cluster: BadRequestException: You can't delete cluster in DELETING state.
+		if strings.Contains(err.Error(), "DELETING") {
+			return nil
+		}
```
Comment (hasheddan) [inline pkg/controller/kafka/cluster/setup.go:74]: @haarchri this is likely a fine solution for now, but does AWS distinguish this as a specific type of error (e.g. is the status code unique). That may be a little more future-proof if this error message changed in the future.

**Review by hasheddan — APPROVED** [review]
@haarchri assuming the `BadRequest` only happen on delete when the cluster is in `DELETING` state, this LGTM! I see you have a new merge commit, could you remove that and instead rebase your changes on `master`? Let me know if you have any questions about how to do so :)

**Comment by haarchri** [comment]
history cleanup done & final test done

aws-console:
image

crossplane:
```
kubectl apply -f cluster.yaml 
cluster.kafka.aws.crossplane.io/kafka-test-server created

NAME                                                READY   SYNCED   EXTERNAL-NAME
cluster.kafka.aws.crossplane.io/kafka-test-server   True    True     arn:aws:kafka:eu-central-1:255932642927:cluster/kafka-test-server/49ad028c-aead-4d05-8519-c725e0828854-1

kubectl delete cluster.kafka.aws.crossplane.io/kafka-test-server
cluster.kafka.aws.crossplane.io "kafka-test-server" deleted
```

File: examples/kafka/cluster.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# It can take approximately 15-25 minutes to create a cluster.
+apiVersion: kafka.aws.crossplane.io/v1alpha1
+kind: Cluster
+metadata:
+  name: kafka-test-server
+spec:
+  forProvider:
+    brokerNodeGroupInfo:
+      clientSubnets:
```
Comment (hasheddan) [inline examples/kafka/cluster.yaml:9]: @haarchri I realized after merge that we can reference subnets in this resource, could you make sure to follow up and add reference fields to `Subnet` for that?

Comment (haarchri) [reply examples/kafka/cluster.yaml:9]: @hasheddan added via https://github.com/crossplane/provider-aws/pull/762 - if you have time would be cool to get a review

**Pull request merged**